web/packages/teleport/src/JoinTokens/CreateJoinTokenDialog.tsx

```diff
@@ -0,0 +1,568 @@
+  border-width: 2px;
+  border-style: solid;
+  border-radius: ${props => props.theme.radii[2]}px;
+
+  margin-bottom: ${props => props.theme.space[3]}px;
+
+  padding-top: ${props => props.theme.space[3]}px;
+  padding-bottom: ${props => props.theme.space[3]}px;
+  padding-left: ${props => props.theme.space[3]}px;
+  padding-right: ${props => props.theme.space[3]}px;
+`;
+
+const JoinTokenIAMForm = ({
+  rules,
+  setTokenState,
+}: {
+  rules: AWSRules[];
+  setTokenState: React.Dispatch<React.SetStateAction<NewJoinTokenState>>;
+}) => {
+  function removeRule(index: number) {
+    const newRules = [...rules].filter((rule, i) => {
+      if (index === i) {
+        return false;
+      }
+      return rule;
+    });
+    setTokenState(prevState => ({
+      ...prevState,
+      iam: newRules,
+    }));
+  }
+
+  function setTokenRulesField(
+    ruleIndex: number,
+    fieldName: string,
+    value: string
+  ) {
+    setTokenState(prevState => ({
+      ...prevState,
+      [prevState.method.value]: prevState[prevState.method.value].map(
+        (rule, i) => {
+          if (ruleIndex !== i) {
+            return rule;
+          }
+          return {
+            ...rule,
+            [fieldName]: value,
+          };
+        }
+      ),
+    }));
+  }
+
+  function addNewRule() {
+    setTokenState(prevState => ({
+      ...prevState,
+      iam: [...prevState.iam, { aws_account: '' }],
+    }));
+  }
+
+  return (
+    <>
+      {rules.map((rule, index) => (
+        <RuleBox
+          key={index} // order does not change without updating the reference array
+        >
+          <Flex alignItems="center" justifyContent="space-between">
+            <Text fontWeight={700} mb={2}>
+              AWS Account
+            </Text>
+
+            <ButtonIcon onClick={() => removeRule(index)}>
+              <Trash size={16} color="text.muted" />
+            </ButtonIcon>
+          </Flex>
+          <FieldInput
+            label="Account ID"
+            rule={requiredField('Account ID is required')}
+            placeholder="aws account ID"
+            value={rule.aws_account}
+            onChange={e =>
+              setTokenRulesField(index, 'aws_account', e.target.value)
+            }
+          />
+          <FieldInput
+            label="ARN"
+            toolTipContent={`The joining nodes must match this ARN. Supports wildcards "*" and "?"`}
+            placeholder="AWS ARN"
+            value={rule.aws_arn}
+            onChange={e => setTokenRulesField(index, 'aws_arn', e.target.value)}
+          />
+        </RuleBox>
+      ))}
+      <ButtonText onClick={addNewRule}>
+        <Plus size={16} mr={2} />
+        Add another AWS Account
+      </ButtonText>
+    </>
+  );
+};
+
+type OptionGCP = Option<string, string>;
+
+const JoinTokenGCPForm = ({
+  rules,
+  setTokenState,
+}: {
+  rules: NewJoinTokenGCPState[];
+  setTokenState: React.Dispatch<React.SetStateAction<NewJoinTokenState>>;
```

Comment:
I've never really been a fan of passing the `React.Dispatch` method to child components - I think components should care about themselves and offer a `onXXX` prop, then the parent component updates its own state instead of passing down the setState method

https://medium.com/@christopherthai/why-you-shouldnt-pass-react-s-setstate-as-a-prop-a-deep-dive-8a3dcd74bec8

Reply:
Great point, will change!